### PR TITLE
Enhance RDF file to include comprehensive repository url processing status

### PR DESCRIPTION
**Pull request by tnzmnjm**
The goal is to produce a .ttl file which has all the information of the `repourl`s. Previously rows with null values, dupliates, or incomplete urls were dropped but in this approach no rows will be dropped and any history of actions on the rows will be captured in a new column(`explanation`).

The other addition is writing the specific data with a specific format in a `JSON Lines` file which will be used for creating a Sankey diagram  @(#56)

The script  `utils/initial_data_preparation.py` will produce a dataframe with columns : `['projectref', 'nlnetpage', 'repourl', 'duplicate_flag', 'null_value_flag', 'repodomain', 'domain_extraction_flag', 'incomplete_url_flag', 'base_repo_url', 'base_repo_url_flag']`

The script `src/github_repo_request_local.py`  will add these 4 columns : `'testfilecountlocal', 'clone_status', 'last_commit_hash', explanation` 

This df is then passed to the function `dataframe_to_ttl` from the script `utils/export_to_rdf.py` which now have all these columns information .

### CODE REVIEW

**Review by julianharty — COMMENTED** [review]
Overall the code's easy to read and the code documentation clear, thank you very much. I've added some suggestions and questions that might further improve the code and the outputs it provides. I'm happy to approve the code once you've had the opportunity to consider and apply which ever of these are appropriate at the moment.

File: src/github_repo_request_local.py
Comment (tnzmnjm) [reply src/github_repo_request_local.py:180]: - Created the script `tests/test_add_explanations.py` and  introduced parameterised tests for this function.

File: src/github_repo_request_local.py
Comment (tnzmnjm) [reply src/github_repo_request_local.py:190]: Defined the constant `EXPECTED_URL_PARTS = 5`  for url validation in scripts `src/github_repo_request_local.py` and `utils/initial_data_preparation.py`

I removed the `rstrip("/")` . But I will need to get the parts here as I'm not saving them anywhere when running the `initial_data_preparations` script.

File: src/github_repo_request_local.py
Comment (tnzmnjm) [reply src/github_repo_request_local.py]: This is done.

File: src/github_repo_request_local.py
Comment (tnzmnjm) [reply src/github_repo_request_local.py:216]: The first item is done.

File: src/github_repo_request_local.py
Comment (tnzmnjm) [reply src/github_repo_request_local.py:287]: Added `# Skip this iteration and move to the next row as the URL lacks specific repository details(owner/reponame)`

File: src/github_repo_request_local.py
Comment (tnzmnjm) [reply src/github_repo_request_local.py:428]: This is done. 
`script_name = os.path.basename(__file__)`
`logger.info(f"Script {script_name} execution completed.")`

File: utils/export_to_rdf.py
Comment (tnzmnjm) [reply utils/export_to_rdf.py:29]: - Descriptions added
- Issue created #71

File: utils/initial_data_preparation.py
Comment (julianharty) [inline utils/initial_data_preparation.py:152]: Here's an example where the flag's name might communicate the intention of the variable (the column) more clearly; and/or there may be a mistake in this line of code.

The scheme couldn't be extracted by our code (line 142 in this file) rather than the domain being extracted. From line 137 (part of the docstring) it says `unsupported_url_flag` columns is returned not `domain_extraction_flag`. Should this code be setting `unsupported_url_flag` instead?

Comment (julianharty) [reply utils/initial_data_preparation.py:152]: PS: I prefer `unsupported_url_scheme` as the column name as I think that is clearer than either of these two columns names.

File: utils/initial_data_preparation.py
Comment (julianharty) [inline utils/initial_data_preparation.py:199]: Are we actually removing rows with incomplete URLs? I thought we're now keeping these rows in the overall dataframe, however the rows are marked as having an incomplete URL for their repo URL.

File: utils/initial_data_preparation.py
Comment (julianharty) [inline utils/initial_data_preparation.py:337]: I find the `mark_...` function names clearer - unless we're actually removing these rows with incomplete URLs from the massive dataframe. How about renaming the function if it's now marking rather than removing rows?

**Review by julianharty — CHANGES_REQUESTED** [review]
Overall really good progress. Let's see if we can wrap up this PR with minimal changes before we merge it. For me, the key potential issue is whether we're filtering out or marking incomplete URLs; and there might be a bug in one of the functions. 

Once you've read my comments pick those that you believe are worth actioning now and we can create issues for anything else we'd like to address in the near future.

File: README.md
Comment (tnzmnjm) [reply README.md]: This is done.

File: README.md
Comment (tnzmnjm) [reply README.md:104]: Sure. Thanks.

File: src/github_repo_request_local.py
Comment (tnzmnjm) [reply src/github_repo_request_local.py:174]: Would you like me to move this section to the script `export_to_rdf.py`?

File: tests/test_add_explanations.py
Comment (tnzmnjm) [reply tests/test_add_explanations.py:1]: Changes have been made on the `add_explanations` function and respective test cases. No change on the matching cases. They are still the same.

File: tests/test_add_explanations.py
Comment (tnzmnjm) [reply tests/test_add_explanations.py:84]: I would like to use `PyHamcres`. Thank you.

File: utils/initial_data_preparation.py
Comment (tnzmnjm) [reply utils/initial_data_preparation.py:28]: I have tried to put  detailed `docstring`s at the beginning of each function. I agree that adding one line summary would help the reader and makes it easier to maintain.

File: utils/initial_data_preparation.py
Comment (tnzmnjm) [reply utils/initial_data_preparation.py:156]: As the name of the function is `extract_and_flag_domains` and in the helper function `get_domain` we are trying to extract the `hostname`, I thought adding `domain_extraction_flag` would be a better choice. However, I agree that `unsupported_url_scheme` is a more precise choice. And I will change the docstring as well.
Thanks for your feedback.

File: utils/initial_data_preparation.py
Comment (tnzmnjm) [reply utils/initial_data_preparation.py:211]: No we do not remove any rows at any points. I can change the function name to `marking_incomplete_urls`. I will need to make small changes in the test file `test_github_repo_request_local.py`.

File: utils/initial_data_preparation.py
Comment (tnzmnjm) [reply utils/initial_data_preparation.py:354]: Yes. I will do this now.

File: utils/initial_data_preparation.py
Comment (tnzmnjm) [reply utils/initial_data_preparation.py]: - Renamed column 'domain_extraction_flag' to 'unsupported_url_scheme' throughout the codebase as per review feedback.
- Updated references and function calls accordingly to reflect the new column name.

File: utils/initial_data_preparation.py
Comment (tnzmnjm) [reply utils/initial_data_preparation.py]: - Function name has been changed across the codebase
- test cases for this function has been modified

File: utils/initial_data_preparation.py
Comment (julianharty) [inline utils/initial_data_preparation.py:245]: Given our discussion about possible bugs in removing '/' characters incorrectly at times, could this be tightened up to remove only the '/' if it is the leading character(s). 

FWIW using the interactive python editor https://www.w3schools.com/python/ref_string_lstrip.asp we can quickly see the effects of the 3 variations of the strip function.

```
txt = "//domain.com/owner/repo/"
print(txt.lstrip('/'))
print(txt.strip('/'))
print(txt.rstrip('/'))
```

**Review by julianharty — APPROVED** [review]
Thank you very much, it's great to see the various improvements in action.

I've approved this and please merge this PR once a couple of points are addressed:

- Reverting the deletion of `data/.gitkeep`
- Adding the version info for `gc` in requirements.txt
- Create tickets for any work to follow up on

And you're welcome to make minor tweaks for any other comments here before merging the PR.

File: src/github_repo_request_local.py
Comment (tnzmnjm) [reply src/github_repo_request_local.py:174]: Created the issue #74

File: src/github_repo_request_local.py
Comment (tnzmnjm) [reply src/github_repo_request_local.py]: Created an issue #75.

File: requirements.txt
Comment (tnzmnjm) [reply requirements.txt:11]: - double checked to see if there is any specific  version for this module
- it is built into the Python runtime environment --> there's no package version related to Python's garbage collection

File: data/.gitkeep
Comment (tnzmnjm) [reply data/.gitkeep:1]: I pushed it back --> `git add ../data/.gitkeep`

File: README.md
Comment (tnzmnjm) [reply README.md]: This is fixed.

File: src/github_repo_request_local.py
Comment (tnzmnjm) [reply src/github_repo_request_local.py:190]: Added this review in the issue #74

File: tests/test_github_repo_request_local.py
Comment (tnzmnjm) [reply tests/test_github_repo_request_local.py:23]: It is a bad URL because of triple slash (///). 
The conversion of `http` to `https` is done in the line 351 in the script `initial_data_preparation.py`. Added a comment to the issue #72 to add test cases for this.

File: src/sankey_diagram_plotly.py
Comment (tnzmnjm) [reply src/sankey_diagram_plotly.py:436]: Added comments and explained this change is part of an ongoing experiment to optimise test runner detection

File: utils/initial_data_preparation.py
Comment (tnzmnjm) [reply utils/initial_data_preparation.py]: This is done.

File: utils/initial_data_preparation.py
Comment (tnzmnjm) [reply utils/initial_data_preparation.py]: This is done.

File: utils/initial_data_preparation.py
Comment (tnzmnjm) [reply utils/initial_data_preparation.py:243]: Given there are `repourls` like `https://github.com/inters/vita/` which have `/` at the beginning and end, `strip` would remove both of them.Please see below:
`parsed_url = urlparse("https://github.com/inters/vita/")`
`path = parsed_url.path.strip("/")`
`parts = path.split("/")` --> ['inters', 'vita']

But if `lstrip` is used: `parts` would become `['inters', 'vita', '']` which then makes the rest of the actions wrong (base_path = "/".join(parts[:2]) )

**Pull request merged**
